src/rgw/driver/rados/rgw_data_sync.cc

```diff
@@ -6079,7 +6079,7 @@ int RGWSyncBucketCR::operate(const DoutPrefixProvider *dpp)
 
         yield call(new RGWSyncBucketShardCR(sc, data_lease_cr, sync_pair,
                                             sync_pipe, bucket_stopped,
-                                            bucket_status.incremental_gen, tn, progress));
+                                            gen ? *gen : bucket_status.incremental_gen, tn, progress));
```

Comment:
@cbodley it's not - actually with this commit (https://github.com/ceph/ceph/commit/a191f9a6d2997f655f580e16731615e77350a03d) it will continue as there is an `else` surrounding it.

Reply:
thanks, that looks like a bug? we probably want a `return set_cr_done()` in the first branch. cc @smanjara

Reply:
that was my confusion, but if we return cr done what would happen to the requested gen? how is that gonna retry later? are we logging the requested gen anywhere?

Reply:
true, that branch is adding an error repo entry for the current_gen, not gen. maybe the error repo entry wasn't redundant after all?

Reply:
perhaps, but why shouldn't we continue with the requested gen? any possible problems?

Reply:
it's a matter of prioritization. we want to focus on finishing the oldest generation first so we can retire it's logs

Reply:
oh probably it was retrying requested gen that I thought was redundant here https://github.com/ceph/ceph/pull/59580?

Reply:
probably, but still, I'm confused as to why the CR can't continue with the request gen and we should log both in the error repo. why can't we utilize the current CR?

Reply:
the idea is to finish processing whatever generation we get from  `bucket_status.incremental_gen` first so that we mark the shards for that generation as done before we move on to the next generation in an order during bucket incremental sync. (its contrary to what I said in the chat. apologies for that) we update the shards as done with a particular generation in the bucket sync status [here](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L5187) 
so any requests that comes with a different gen should go to the error repo which would correctly happen if we reverted https://github.com/ceph/ceph/pull/59580.
as for [this](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L6035-L6049) block, I think there was a particular edge case for the `current_gen` we were trying to work around. I do think, as Casey said, should return set_cr_done() after writing to error repo there.

Reply:
I see, thanks for the explanation. but so we stuck the replication until that generation is proceeding now if I understand correctly.

> as for [this](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L6035-L6049) block, I think there was a particular edge case for the current_gen we were trying to work around. I do think, as Casey said, should return set_cr_done() after writing to error repo there.

If we return `set_cr_done()` then the `gen` would be lost I believe - basically this was my test case. based on your explanation, I guess reverting that PR should be enough, or what was the edge case?

Reply:
in the comment [here](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L6036), it says the error repo retry was added due to a sync issue when data log entry is missing. I am not exactly sure how that happens.